spark/spark-tensorflow-distributor/spark_tensorflow_distributor/mirrored_strategy_runner.py

```diff
@@ -0,0 +1,311 @@
+                        'were available.'
+                    )
+                else:
+                    os.environ['CUDA_VISIBLE_DEVICES'] = ','.join(str(e) for e in range(self.num_processors))
+            elif self.processor_type == 'cpu':
+                if self.num_processors > 1:
+                    raise ValueError(
+                        f'Cannot run with more than 1 CPU machine in local mode. '
+                        'Try setting num_processors to -1.'
+                    )
+                os.environ['CUDA_VISIBLE_DEVICES'] = ''
+            return MirroredStrategyRunner.run_tensorflow_program(train_fn, self.use_custom_strategy, **kwargs)
+
+        # Run in distributed mode
+        self._check_encryption()
+        self.logger.info('Distributed training in progress...')
+        self.logger.info('View Spark worker logs to inspect training...')
+        spark_output = self.sc.parallelize(range(self.num_workers), self.num_workers) \
+            .barrier() \
+            .mapPartitions(spark_task_program) \
+            .collect()
+        training_info, results = zip(*[e for e in spark_output if e is not None])
+        num_processors_used_for_training = training_info[0]
+        self.logger.info(
+            f'Training is complete! Used {num_processors_used_for_training} processors of type `{self.processor_type}`.'
+        )
+        return list(results)
+
+    # Gets the number of GPUs assigned to each task
+    # Trains on as few workers as possible while still having an even distribution
+    # of GPUs and a total number of GPUs specified by the user
+    @staticmethod
+    def _get_gpu_allocation(num_processors, num_gpus_owned_by_tasks):
+        if not all(e == num_gpus_owned_by_tasks[0] for e in num_gpus_owned_by_tasks):
+            raise Exception(
+                'MirroredStrategyRunner does not support workers with different numbers of GPUs.'
+            )
+        pid_of_last_task_used = 0
+        num_gpus_owned_by_used_tasks = 0
+        for i in range(len(num_gpus_owned_by_tasks)):
+            num_gpus_owned_by_used_tasks += num_gpus_owned_by_tasks[i]
+            pid_of_last_task_used = i
+            if num_gpus_owned_by_used_tasks >= num_processors:
+                break
+        gpus_to_allocate = [
+            e for e in num_gpus_owned_by_tasks[:pid_of_last_task_used + 1]
+        ] + [
+            0 for e in num_gpus_owned_by_tasks[pid_of_last_task_used + 1:]
+        ]
+        num_gpus_available = sum(gpus_to_allocate)
+        num_gpus_to_remove = num_gpus_available - num_processors
+        num_gpus_to_remove_per_task = num_gpus_to_remove // (pid_of_last_task_used + 1)
+        for i in range(pid_of_last_task_used + 1):
+            gpus_to_allocate[pid_of_last_task_used - i] -= num_gpus_to_remove_per_task
+        for i in range(num_gpus_to_remove - num_gpus_to_remove_per_task * (pid_of_last_task_used + 1)):
+            gpus_to_allocate[pid_of_last_task_used - i] -= 1
```

Comment:
Is this to calculate the overflow? If so, you could do something like 


```suggestion
        extra_gpus_to_remove = num_gpus_to_remove % (pid_of_last_task_used + 1)
        for i in range(extra_gpus_to_remove): 
            gpus_to_allocate[pid_of_last_task_used - i] -= 1
```